### PR TITLE
Only org policy

### PR DESCRIPTION
**Pull request by addisonbeck**
## Narrative  ✍🏻
As an enterprise organization admin I want to be able to block members of my organization from being able to be a part of other organizations.

## Acceptance Criteria Relevant To This PR 💯
1. A policy (OnlyOrg) exists that restricts non-admin members from being able to be a part of other organizations.
2. When applying the OnlyOrg policy any non-owner members who aren't the one enabling the policy & that are already in another org should be removed from the org applying the policy. This user flow should mirror what already exists for the 2FA policy.
3. Removed users should get an email informing them of the removal.
4. The OnlyOrg policy needs to be configurable from the web vault AND the enterprise portal.
5. Users in an org with the OnlyOrg policy enabled should not be able to create new organizations.

### API

- [x] Added a new PolicyType enum
- [x] Blocked accepting new organization invites when OnlyOrg is relevant to the user
- [x] Blocked creating new orgs if already in an org with OnlyOrg enabled (this is also checked on the client)
- [x] Blocked confirming any org members that may have violated OnlyOrg between accepting their invite and being confirmed
- [x] Created an email alert to be sent when a user is kicked because of the OnlyOrg policy
- [x] Removed users and sent alerts when appropriate (this involved a slight refactor of the same logic for 2FA)

### Portal

- [x] Added any needed localization strings for policy config in the portal
- [x] Allowed for the OnlyOrg policy to be configured on the policies screen

## Screenshots
Screen Shot 2020-10-08 at 1 29 23 PM
Screen Shot 2020-10-08 at 1 29 29 PM

## Notes 📝
[Asana ticket](https://app.asana.com/0/1195018406457675/1196488889925152)
[Web PR](https://github.com/bitwarden/web/pull/669)
[jslib PR](https://github.com/bitwarden/jslib/pull/183)

### CODE REVIEW

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (kspearrin) [inline src/Core/Services/Implementations/OrganizationService.cs:1121]: You cannot join this organization while you are a member of another organization.

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (kspearrin) [inline src/Core/Services/Implementations/OrganizationService.cs:1128]: member of.

File: src/Core/Services/Implementations/PolicyService.cs
Comment (kspearrin) [inline src/Core/Services/Implementations/PolicyService.cs:57]: formatting here. I think you need to drop the `()`

File: src/Core/Services/Implementations/PolicyService.cs
Comment (kspearrin) [inline src/Core/Services/Implementations/PolicyService.cs:90]: same. i think you need to check status

File: src/Core/Services/Implementations/PolicyService.cs
Comment (kspearrin) [inline src/Core/Services/Implementations/PolicyService.cs:100]: Seems like this operation could get expensive in a loop of possibly thousands of users. Any way to optimize further?

**Review by kspearrin — CHANGES_REQUESTED** [review]

File: src/Core/Services/Implementations/PolicyService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs:100]: For sure. Setting up some kind of bulk GET stored procedure that accepts a list of orgUser Ids and gets all of those at once would cut trips to the DB for that operation down to one, but there doesn't seem to be a precedent for a "get by many" type of procedure so far, so let me know if that's the direction you'd like to go before I make the swap, @kspearrin .

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs:1113]: `Lazy<T>` doesn't seem to support async operations. I found a few blogs and the like suggesting things like [creating a class and manually managing the tasks](https://devblogs.microsoft.com/pfxteam/asynclazyt/), but I'm not sure if that level of investment is worth it here, since it would ultimately result in more lines of code than just checking for a null value. I'm not very familiar with the Lazy<T> class, though (thanks for the TIL), so maybe I'm just missing something.

Another option would be to do something like this:

```
ICollection<Policy> orgPolicies = null;
var userOrgs = await _organizationUserRepository.GetManyByUserAsync(user.Id);
var checkOnlyOrg = userOrgs.Any(ou => ou.OrganizationId != orgUser.OrganizationId && ou.Status != OrganizationUserStatusType.Invited);
var checkTwoFactor = !await userService.TwoFactorIsEnabledAsync(user);
if (checkOnlyOrg || checkTwoFactor)
{
    orgPolicies = await _policyRepository.GetManyByOrganizationIdAsync(orgUser.OrganizationId);
};

if (checkOnlyOrg)
{   
    if (orgPolicies.Any(policy => policy.Type == PolicyType.OnlyOrg && policy.Enabled))
    {
        throw new BadRequestException("You cannot join this organization until you are not a " +
            "part of any other organizations.");
    }
    var userPolicies = await _policyRepository.GetManyByUserIdAsync(user.Id);
    if (userPolicies.Any(policy => policy.Type == PolicyType.OnlyOrg))
    {
        throw new BadRequestException("You cannot join this organization because you are already " +
            "member of another organization that restricts having multiple organizations.");
    }
}

if (checkTwoFactor)
{
    if (orgPolicies.Any(p => p.Type == PolicyType.TwoFactorAuthentication && p.Enabled))
    {
        throw new BadRequestException("You cannot join this organization until you enable " +
            "two-step login on your user account.");
    }
}
```

File: bitwarden_license/src/Portal/Models/PolicyEditModel.cs
Comment (addisonbeck) [reply bitwarden_license/src/Portal/Models/PolicyEditModel.cs]: The 2FA policy wasn't represented in this switch at all, which was causing a bug: enabling the 2FA policy from the portal simply didn't work.

I noticed this, but it didn't seem in scope of this PR so I was planning to fast-follow and fix it later, like with deleting orgUser records with an org, but since you brought it up I went ahead and fixed it here 😁

File: src/Api/Controllers/OrganizationsController.cs
Comment (addisonbeck) [reply src/Api/Controllers/OrganizationsController.cs]: done

File: src/Api/Controllers/OrganizationsController.cs
Comment (addisonbeck) [reply src/Api/Controllers/OrganizationsController.cs]: done

File: src/Core/MailTemplates/Handlebars/OrganizationUserRemovedForPolicyOnlyOrg.html.hbs
Comment (addisonbeck) [reply src/Core/MailTemplates/Handlebars/OrganizationUserRemovedForPolicyOnlyOrg.html.hbs]: done

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: done

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: Good catch

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: I think you meant status != invited, right? I added a check, but did it for Invited since that seemed like what made sense.

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs:1121]: Done

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: done

File: src/Core/Services/Implementations/PolicyService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs]: done

File: src/Core/Services/Implementations/PolicyService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs]: checked

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1121]: Consider rewording:
```
"You may not join this organization until you leave or remove all other organizations."
```

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1127]: Consider rewording:
```
"You may not join this organization because you are a member of an organization which forbids it."
```

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (cscharf) [reply src/Core/Services/Implementations/OrganizationService.cs:1113]: that could work, or just a lightweight lazyish implementation with an in-method function:

```c#
// inside method ...
ICollection<Policy> orgPolicies = null;
async Task<bool> hasPolicyAsync(PolicyType policyType)
{
    if (orgPolicies == null)
    {
        orgPolicies = await _policyRepository.GetManyByOrganizationIdAsync(orgUser.OrganizationId);
    }
    return orgPolicies.Any(p => p.Type == policyType && p.Enabled);
}

// ... more fun stuff
if (userOrgs.Any(ou => ou.OrganizationId != orgUser.OrganizationId && ou.Status != OrganizationUserStatusType.Invited))
{
    if (await hasPolicyAsync(PolicyType.OnlyOrg))
    {
        throw new BadRequestException("...");
    }
}

// ... etc.
```

File: src/Core/Services/Implementations/PolicyService.cs
Comment (cscharf) [inline src/Core/Services/Implementations/PolicyService.cs:70]: Would we not kick out the user setting the policy if they are not the owner BUT are in violation? Seems silly but not sure what's appropriate here. Will need @kspearrin to weight in.

Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs:70]: This was actually a check I just moved up from [line 65](https://github.com/bitwarden/server/pull/962/files/de295deb245da6ebd54e1b32c7c12b58ac9adb62#diff-928f61507e60ccef15abcc2ae6b96ae6125cd05a80e6ef2f34439f7914c29b36L65), so it is something we were already filtering on for 2FA.

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: changed

File: src/Core/Enums/PolicyType.cs
Comment (addisonbeck) [reply src/Core/Enums/PolicyType.cs]: added

File: src/Api/Controllers/OrganizationsController.cs
Comment (addisonbeck) [reply src/Api/Controllers/OrganizationsController.cs]: done

File: src/Api/Controllers/OrganizationsController.cs
Comment (addisonbeck) [reply src/Api/Controllers/OrganizationsController.cs]: done

File: bitwarden_license/src/Portal/Models/PolicyEditModel.cs
Comment (addisonbeck) [reply bitwarden_license/src/Portal/Models/PolicyEditModel.cs]: done

File: src/Api/Controllers/OrganizationsController.cs
Comment (addisonbeck) [reply src/Api/Controllers/OrganizationsController.cs]: good eye 👀

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs:1113]: @cscharf that's sleek. I implemented based on your suggested, but added some fluff to it so I could also check userPolicies with the same method.

**Comment by addisonbeck** [comment]
@cscharf @kspearrin this one got a little muddy with the string changes (I'll try and run those by you guys before PR next time), but I think the only issue left open is [this one](https://github.com/bitwarden/server/pull/962#discussion_r502380882) regarding optimizing removal of users on policy enable. Let me know if you'd like me to make the change I suggested on that thread, or any others.

File: src/Core/Services/Implementations/PolicyService.cs
Comment (cscharf) [reply src/Core/Services/Implementations/PolicyService.cs]: @addisonbeck , there's a UDT in the database, `GuidIdArray` which is basically a `TABLE` type with a single column, `[Id] UNIQUEIDENTIFIER NOT NULL` that you an use for creating a bulk stored procedure that takes in an `IEnumerable<Guid>` and returns the data in a set operation (`JOIN @tableVariable AS Ids ON -- etc...`).

See `[dbo].[Cipher_UpdateCollections]` for a "kind of" example where this is used (at least as a sproc parameter and Dapper mapper, the use-case it a bit difference once you have it in the sproc).

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1124]: Not that you _should_, but could make this a 1-liner 😁 :
```c#
policies = userPolicies = userPolicies ?? await _policyRepository.GetManyByUserIdAsync(user.Id);
```

Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: Done + reduced line count even more by using a ternary

File: src/Core/Services/Implementations/PolicyService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs]: Done!

File: src/Sql/dbo/Stored Procedures/OrganizationUser_ReadByUserIds.sql
Comment (cscharf) [inline src/Sql/dbo/Stored Procedures/OrganizationUser_ReadByUserIds.sql:12]: Reduce indentation here to match the rest of the sproc

File: src/Sql/dbo/Stored Procedures/OrganizationUser_ReadByUserIds.sql
Comment (cscharf) [inline src/Sql/dbo/Stored Procedures/OrganizationUser_ReadByUserIds.sql:17]:
```sql
[UserId] IN (SELECT [Id] FROM @UserIds)
```

File: util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql
Comment (cscharf) [inline util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql:23]:
```sql
[UserId] IN (SELECT [Id] FROM @UserIds)
```

File: util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql
Comment (cscharf) [inline util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql:18]: Fix indentation here as well

File: util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql
Comment (cscharf) [inline util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql:24]: Terminate script with a final `GO` statement line

File: src/Sql/dbo/Stored Procedures/OrganizationUser_ReadByUserIds.sql
Comment (addisonbeck) [reply src/Sql/dbo/Stored Procedures/OrganizationUser_ReadByUserIds.sql]: This one was strange, as that line was indented as expected locally and git wasn't showing a change. I tabbed+commit, then untabbed+commited and that seemed to resolve. Should be fixed.

File: util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql
Comment (addisonbeck) [reply util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql]: Done

File: util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql
Comment (addisonbeck) [reply util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql]: Done

File: util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql
Comment (addisonbeck) [reply util/Migrator/DbScripts/2020-10-14_00_OrgUserReadByUserIds.sql:24]: GO'd

**Review by cscharf — APPROVED** [review]
Will still need @kspearrin to review from his requested changes and approval before this can be merged.

File: src/Core/Services/Implementations/PolicyService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs:82]: userOrgs = all the dbo.organizationUser records for the current person the loop is iterating on
removableOrgUsers = the people we will want to remove from the organization applying the policy

One is orgUser records for an organization, the other is orgUser records for a given user inside that org

File: src/Core/Services/Implementations/OrganizationService.cs
Comment (addisonbeck) [reply src/Core/Services/Implementations/OrganizationService.cs]: Was this comment looking at a specific commit? This logic was moved around on [this PR](https://github.com/bitwarden/server/pull/945/files), it's reflected in some commits here because of merge issues.

File: src/Core/Services/Implementations/PolicyService.cs
Comment (kspearrin) [reply src/Core/Services/Implementations/PolicyService.cs:82]: Right, but instead of querying the database again, couldn't you just run a in-memory query on the existing `orgUsers` that you already pulled earlier?

Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs:82]: No, because orgUsers is filtered on an organizationId - the id for the organization applying the policy. The logic that uses userOrgs needs all organizationUser records for that user, regardless of organization. @kspearrin

**Review by kspearrin — APPROVED** [review]

**Pull request merged**
